### PR TITLE
Alternative improved rune recognition

### PR DESCRIPTION
**Pull request by bsxf-47**
The new rune recognition algorithm 

Features:
- can insert new definitions by simply adding a new element
- no initialization needed
- definitions use direction numbers of the old recognition
- disabled by default, can be enabled in cfg.ini, option `use_alt_rune_recognition` in `[input]`

Performance impact: negligible, according to my measurements, 98% of the time is spent inside handleRuneDetection(), which is beyond the scope of the code.

Any suggestions how to make it better are very welcome.

### CODE REVIEW

File: src/game/Spells.cpp [hunk 1]
```diff
@@ -632,8 +632,10 @@ void ARX_SPELLS_ManageMagic() {
 			ARX_SPELLS_Analyse();
 			if(!SpellMoves.empty())
 				ARX_SPELLS_AnalyseSYMBOL();
+		} else {
+			SpellRecognition::ARX_SPELLS_Analyse_DTW();
 		}
-
```
Comment (Eli2) [inline src/game/Spells.cpp]: Unrelated whitespace change

File: src/game/magic/SpellRecognition.cpp [hunk 2]
```diff
@@ -35,9 +36,455 @@
...
+/*!
+* Resample the input while keeping key points contained in indices
+*/
+void resample(std::vector<Vec2f> &in, std::vector<int> &indices, std::vector<Vec2f> &out) {
+	
+	float totalLen = 0.0;
+	{
+		//find the total length
+		Vec2f p1 = in[0];
+		Vec2f p2;
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Move to smaller scope

File: src/game/magic/SpellRecognition.cpp [hunk 3]
```diff
@@ -35,9 +36,455 @@
...
+
+/*!
+ * Convert a vector of input points to a sequence of directions
+ */
+void patternToDirs(std::vector<Vec2f> &in, int *out) {
+	Vec2f thisPoint, prevPoint;
+	prevPoint = in[0];
+	for(int i = 1; i < (int)in.size(); i++) {
+		thisPoint = in[i];
+		//out[i - 1] = quantize(angle(thisPoint - prevPoint));
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Remove

File: src/game/magic/SpellRecognition.cpp [hunk 4]
```diff
@@ -35,9 +36,455 @@
...
+	}
+
+	return angle;
+}
+
+/*!
+ * Convert a vector of input points to a sequence of directions
+ */
+void patternToDirs(std::vector<Vec2f> &in, int *out) {
+	Vec2f thisPoint, prevPoint;
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: thisPoint to smaller scope, declare prevPoint in the next line

File: src/game/magic/SpellRecognition.cpp [hunk 5]
```diff
@@ -35,9 +36,455 @@
...
+	int minX = plist[0].x;
+	int minY = plist[0].y;
+	float currTolerance = 0;
+	for(int i = 1; i < inputSize; i++) {
+		maxX = glm::max(maxX, (int)plist[i].x);
+		maxY = glm::max(maxY, (int)plist[i].y);
+		minX = glm::min(minX, (int)plist[i].x);
+		minY = glm::min(minY, (int)plist[i].y);
+	}
+	currTolerance = ((maxX - minX + maxY - minY) / 2.0f)*TOLERANCE;
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Code style

File: src/game/magic/SpellRecognition.cpp [hunk 6]
```diff
@@ -35,9 +36,455 @@
...
+
+	out.push_back(0);
+	Vec2f lastImp = in.front();
+
+	float minAngle = (2.0 / 3.0)*glm::pi<float>();
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Code style

File: src/game/magic/SpellRecognition.cpp [hunk 7]
```diff
@@ -194,7 +643,7 @@ void ARX_SPELLS_AddPoint(const Vec2s & pos) {
 		return;
 	}
 
-	if(plist.size() == MAX_POINTS) {
+	if((long)plist.size() == MAX_POINTS) {
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Use size_t for MAX_POINTS in a separate commit

File: src/game/magic/SpellRecognition.cpp [hunk 8]
```diff
@@ -35,9 +36,455 @@
...
+				prevPoint = p;
+				newPointAdded = true;
+				continue;
+			} else {
+				D += d;
+			}
+			
+			prevPoint = thisPoint;
+		}
+		if(reallyAdded == (pointsToAdd)) {
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Code style

File: src/game/magic/SpellRecognition.cpp [hunk 9]
```diff
@@ -35,9 +36,455 @@
...
+
+/*!
+ * Resample a direction string to an array of directions
+ */
+void resample(const std::string &in, int *out) {
+
+	int inputSize = in.size();
+	int outputIndex = 0;
+
+	float pointsPerSegment = (float)(directionCount - inputSize) / (float)inputSize;
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: replace c-style cast with constructor style float(...)

**Comment by bsxf-47** [comment]
Updated the code with suggested fixes.

File: src/game/magic/SpellRecognition.cpp [hunk 10]
```diff
@@ -35,9 +36,442 @@
...
+						index++;
+					}
+				}
+			} else {
+				newPointAdded = false;
+			}
+			
+			float d = glm::distance(prevPoint, thisPoint);
+			if((D + d) >= interval) {
+				Vec2f p;
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Why is this done component wise ?
Can ((interval - D) / d) be precalculated ?

File: src/game/magic/SpellRecognition.cpp [hunk 11]
```diff
@@ -35,9 +36,442 @@
...
+ * Inserts indices pointing to the in vector into the out vector.
+ */
+void RuneRecognitionDTW::findKeyPoints(std::vector<Vec2f> &in) {
+	
+	const float TOLERANCE = 0.20f;
+	
+	int inputSize = in.size();
+	
+	//calculate tolerance based on the overall size of the drawing
+	float maxX = in[0].x;
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Why not use Vec2f here ?

File: src/game/magic/SpellRecognition.cpp [hunk 12]
```diff
@@ -35,9 +36,442 @@
...
+
+class RuneRecognitionDTW {
+
+static const size_t requiredPointCount = 20;
+static const size_t directionCount = requiredPointCount - 1;
+static const size_t patternCount = ARRAY_SIZE(patternData);
+static const size_t maxDist = 3; //maximum distance from the diagonal the path can stray
+static const int maxTolerance = 3;
+
+int patternDirections[patternCount][directionCount];
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: m_ prefix

File: src/game/magic/SpellRecognition.cpp [hunk 13]
```diff
@@ -35,9 +36,442 @@
...
+	if(outputIndex == directionCount - 1) {
+		out[outputIndex] = convertDirection(in[inputSize - 1]);
+	}
+}
+
+void RuneRecognitionDTW::init() {
+	
+	std::fill_n(accMatrix[0], directionCount * directionCount, std::numeric_limits<int>::max());
+	
+	for(size_t i=0; i < patternCount; i++) {
```
Comment (Eli2) [inline src/game/magic/SpellRecognition.cpp]: Code style

**Comment by Eli2** [comment]
merged
